exporter/signalfxexporter/metricdata_to_signalfxv2.go

```diff
@@ -0,0 +1,403 @@
+	// handy because the SignalFx protobuf needs pointers not constants.
+	// The mapping of these comes from SignalFx docs regarding Prometheus metrics.
+	// https://docs.signalfx.com/en/latest/integrations/agent/monitors/prometheus-exporter.html#overview
+	bucketMetricType     = &sfxMetricTypeCumulativeCounter
+	quantileMetricType   = &sfxMetricTypeGauge
+	sumMetricType        = &sfxMetricTypeCumulativeCounter
+	totalCountMetricType = &sfxMetricTypeCumulativeCounter
+
+	// Some standard dimension keys.
+	// upper bound dimension key for histogram buckets.
+	upperBoundDimensionKey = "upper_bound"
+	// quantile dimension key for summary quantiles.
+	quantileDimensionKey = "quantile"
+
+	// infinity bound dimension value is used on all histograms.
+	infinityBoundSFxDimValue = float64ToDimValue(math.Inf(1))
+)
+
+func metricDataToSingalFxV2(
+	logger *zap.Logger,
+	md consumerdata.MetricsData,
+) (sfxDataPoints []*sfxpb.DataPoint, numDroppedTimeSeries int, err error) {
+
+	// The final number of data points is likely larger than len(md.Metrics)
+	// but at least that is expected.
+	sfxDataPoints = make([]*sfxpb.DataPoint, 0, len(md.Metrics))
+
+	// Labels from Node and Resource.
+	// TODO: Options to add lib, service name, etc as dimensions?
+	//  Q.: what about resource type?
+	nodeAttribs := md.Node.GetAttributes()
+	resourceAttribs := md.Resource.GetLabels()
+	numExtraDimensions := len(nodeAttribs) + len(resourceAttribs)
+	var extraDimensions []*sfxpb.Dimension
+	if numExtraDimensions > 0 {
+		extraDimensions = make([]*sfxpb.Dimension, 0, numExtraDimensions)
+		extraDimensions = appendAttributesToDimensions(extraDimensions, nodeAttribs)
+		extraDimensions = appendAttributesToDimensions(extraDimensions, resourceAttribs)
+	}
+
+	for _, metric := range md.Metrics {
+		if metric == nil || metric.MetricDescriptor == nil {
+			logger.Warn("Received nil metrics data or nil descriptor for metrics")
+			numDroppedTimeSeries += len(metric.GetTimeseries())
+			continue
+		}
+
+		// Build the fixed parts for this metrics from the descriptor.
+		descriptor := metric.MetricDescriptor
+		metricName := &descriptor.Name
+
+		metricType := fromOCMetricDescriptorToMetricType(descriptor.Type)
+		numLabels := len(descriptor.LabelKeys)
+		filteredLabelKeys := make([]*string, numLabels)
+		for i := 0; i < numLabels; i++ {
+			key := filterKeyChars(descriptor.LabelKeys[i].Key)
+			filteredLabelKeys[i] = &key
+		}
+
+		for _, series := range metric.Timeseries {
+			dimensions := make([]*sfxpb.Dimension, numLabels+numExtraDimensions)
+			copy(dimensions, extraDimensions)
+			for i := 0; i < numLabels; i++ {
+				dimension := &sfxpb.Dimension{
+					Key:   filteredLabelKeys[i],
+					Value: &series.LabelValues[i].Value,
+				}
+				dimensions[numExtraDimensions+i] = dimension
+			}
+
+			for _, dp := range series.Points {
+
+				var msec int64
+				if dp.Timestamp != nil {
+					msec = dp.Timestamp.Seconds*1e3 + int64(dp.Timestamp.Nanos)/1e6
```

Comment:
Might be useful as a function. We should move ptime package from Omniion distribution to upstream at some point and add a `Milliseconds(t ptypes.Timestamp) int64` to it. https://github.com/Omnition/omnition-opentelemetry-collector/tree/master/ptypes/ptime

Reply:
I agree but this package doesn't seem to be the ideal location for this function, let's address it when we move the ptime package.